text/0085-source-hash-sourcemap-keying.md

```diff
@@ -0,0 +1,162 @@
+purely by that content hash in recent versions.
+
+The hashes are available via the `Error.prepareStackTrace` API by invoking
+the `getScriptHash` method on a call site object.  There is no fallback
+possible for browsers not supporting script hashes.
+
+The script hashes are documented to be SHA-256.
+
+# Implementation
+
+To support this the changes in our pipeline are limited.
+
+## Protocol Changes
+
+The `sourcemap` image type gains an additional key called `hashes` which
+holds an object with hashes where each key corresponds to the name of a
+hash algorithm and the value to the checksum in a format appropriate for
+that hash.  Today only `sha256` is specified where the value is a SHA-256
+hash in hexadecimal format:
+
+```json
+{
+  "images": [
+    {
+      "type": "sourcemap",
+      "code_file": "http://example.com/file.min.js",
+      "hashes": {
+        "sha256": "2d711642b726b04401627ca9fbac32f5c8530fb1903cc4db02258717921a4881"
+      }
+    }
+  ]
+}
+```
+
+## SDK Changes
+
+On the SDK side we need to capture the content hashes of the referenced
+frames and build up a map over time.  The easiest solution for this is to
+maintain a growing list of hashes and to rely on `Error.prepareStackTrace`
+to capture this:
+
+```javascript
+const scriptHashes = {};
+const originalPrepareStackTrace = Error.prepareStackTrace;
+
+// we use `Error.prepareStackTrace` exclusively for capturing the script hashes
+// and we rely on `err.stack` parsing for the actual stack trace.  As documented
+// getScriptHash() always returns a SHA-256 hash.
+Error.prepareStackTrace = (err, callsites) => {
+  for (var callsite of callsites) {
+    if (!callsite.getScriptHash) {
+      const filename = callsite.getFileName();
+      if (!scriptHashes[filename]) {
+        scriptHashes[filename] = callsite.getScriptHash();
+      }
+    }
+  }
+  return originalPrepareStackTrace ? originalPrepareStackTrace(err, callsites) : err.stack;
+};
+```
+
+The SDK can then feed the script hashes into the sent image list as specified
+above.
+
+## Toolchain / Source Map Changes
+
+For establishing the mapping from minified JavaScript to the source map, the
+source map needs to gain a content hash entry.  As there is currently no
+specification for how that mapping is supposed to be established, we would
+reserve `x-sentry-file-hash-sha256` (header) and `x_sentry_fileHashSha256` (key):
+
+```json
+{
+  "file": "foo.min.js",
+  "x_sentry_fileHashSha256": "2d711642b726b04401627ca9fbac32f5c8530fb1903cc4db02258717921a4881"
+}
+```
+
+Note that they use of the embedded key for now is likely discouraged for the
+made up header as we do not plan on standardizing the file hashes.
+
+To avoid re-calculating the SHA 256 hashes of the minified files constantly we
+also want to add them to the manifest.  For this we could leverage the digest
+header, with the small suboptimal situation that the HTTP digest spec uses
+a base64 hash:
```

Comment:
We should just embrace the reality that the `headers` have nothing to do with HTTP anymore (even though they originally were inspired by that).

This is all a bit bikesheddy, but we should strive for consistency here between the protocol, the sourcemap JSON, and these `headers`. We use a mix of different names and string/objects all over the place.
I believe we are currently limited by `headers` being a `String->String` mapping. But we could relax that as well if needed.

Reply:
I would honestly just move all this stuff out of `headers` and keep toplevel, but as it stands these are actually still http headers in the current implementation. It really just overlays the HTTP fetch.